plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/DockerInstanceRuntimeInfo.java

```diff
@@ -310,16 +316,33 @@ public String projectsRoot() {
     protected Map<String, ServerImpl> getServersWithFilledPorts(final String externalHostame, final String internalHostname, final Map<String, List<PortBinding>> exposedPorts) {
         final HashMap<String, ServerImpl> servers = new LinkedHashMap<>();
 
+        boolean useMappedPorts = true;
+        if (useInternalContainerAddresses()) {
+            if (info.getNetworkSettings() != null && internalHostname.equals(info.getNetworkSettings().getIpAddress())) {
+                useMappedPorts = false;
+            }
+        }
+
         for (Map.Entry<String, List<PortBinding>> portEntry : exposedPorts.entrySet()) {
             // in form 1234/tcp
             String portProtocol = portEntry.getKey();
             // we are assigning ports automatically, so have 1 to 1 binding (at least per protocol)
             String externalPort = portEntry.getValue().get(0).getHostPort();
+
+            // If we are sending messages to the container directly, we don't want to use the mapped ports.
+            String internalHostnameAndPort;
+            if (useMappedPorts) {
+                internalHostnameAndPort = internalHostname + ":" + externalPort;
+            } else {
+                String internalPort = portEntry.getKey().split("/")[0];
```

Comment:
It is better to use portProtocol variable here  instead of **portEntry.getKey()**